gapis/api/vulkan/set_primitive_count_to_one.go

```diff
@@ -0,0 +1,143 @@
+	cmdPipelineBindings := make(map[VkCommandBuffer]VkPipeline)
+
+	return transform.Transform("setPrimitiveCountToOne", func(ctx context.Context,
+		id api.CmdID, cmd api.Cmd, out transform.Writer) {
+
+		s := out.State()
+		l := s.MemoryLayout
+		cb := CommandBuilder{Thread: cmd.Thread(), Arena: s.Arena}
+
+		isIndirectDraw := false
+		var cmdBuf VkCommandBuffer
+		cmdInstanceCount := instanceCount
+
+		switch cmd := cmd.(type) {
+		case *VkCreateGraphicsPipelines:
+			out.MutateAndWrite(ctx, id, cmd)
+			cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+
+			createInfoCount := uint64(cmd.CreateInfoCount())
+			createInfos := cmd.PCreateInfos().Slice(0, createInfoCount, l).MustRead(ctx, cmd, s, nil)
+			pipelineHandles := cmd.PPipelines().Slice(0, createInfoCount, l).MustRead(ctx, cmd, s, nil)
+
+			for i := uint64(0); i < createInfoCount; i++ {
+				primitive := createInfos[i].PInputAssemblyState().MustRead(ctx, cmd, s, nil).Topology()
+				perPipelineVertexCount[pipelineHandles[i]] = getMinimumVertexCountForPrimitive(ctx, primitive)
+			}
+		case *VkCmdBindPipeline:
+			out.MutateAndWrite(ctx, id, cmd)
+			if cmd.PipelineBindPoint() == VkPipelineBindPoint_VK_PIPELINE_BIND_POINT_GRAPHICS {
+				cmdPipelineBindings[cmd.commandBuffer] = cmd.Pipeline()
+			}
+		case *VkCmdDraw:
+			vertexCount := perPipelineVertexCount[cmdPipelineBindings[cmd.commandBuffer]]
+			newCmd := cb.VkCmdDraw(
+				cmd.commandBuffer,
+				u32.Min(vertexCount, cmd.VertexCount()),
+				u32.Min(instanceCount, cmd.InstanceCount()),
+				cmd.FirstVertex(),
+				cmd.FirstInstance(),
+			)
+			out.MutateAndWrite(ctx, id, newCmd)
+		case *VkCmdDrawIndexed:
+			vertexCount := perPipelineVertexCount[cmdPipelineBindings[cmd.commandBuffer]]
+			newCmd := cb.VkCmdDrawIndexed(
+				cmd.commandBuffer,
+				u32.Min(vertexCount, cmd.IndexCount()),
+				u32.Min(instanceCount, cmd.InstanceCount()),
+				cmd.FirstIndex(),
+				cmd.VertexOffset(),
+				cmd.FirstInstance(),
+			)
+			out.MutateAndWrite(ctx, id, newCmd)
+		case *VkCmdDrawIndirect:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+			cmdInstanceCount = cmd.DrawCount()
+		case *VkCmdDrawIndexedIndirect:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+			cmdInstanceCount = cmd.DrawCount()
+		case *VkCmdDrawIndirectCountKHR:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+		case *VkCmdDrawIndexedIndirectCountKHR:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+		case *VkCmdDrawIndirectCountAMD:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+		case *VkCmdDrawIndexedIndirectCountAMD:
+			isIndirectDraw = true
+			cmdBuf = cmd.commandBuffer
+		default:
+			out.MutateAndWrite(ctx, id, cmd)
+			return
+		}
+
+		// Replace indirect draw calls with direct ones.
+		// TODO: Replace with appropriate indirect calls instead.
+		if isIndirectDraw {
+			vertexCount := perPipelineVertexCount[cmdPipelineBindings[cmdBuf]]
+			newCmd := cb.VkCmdDraw(
+				cmdBuf,
+				vertexCount,
```

Comment:
This should also be "min(vertexCount, ...)" since it is legal for the original drawindirect to be 0.

Reply:
I was gonna do that, but those values are stored in a vkbuffer, and I have found no clear way to read back vkbuffers. It's the same reason all indirect calls were replaced with this direct one.

If there is a way to read vkbuffers I'll change it. Otherwise an option would be to just set the vertex count to 0 for all indirect calls. At least that way the calls will still always be valid.

Reply:
There are WAYS to read back vkBuffers, but we certainly don't want ot go down that path, it is horribly complicated. We should then set them to 0. I think we should avoid doing anything that may cause invalid behavior, since it has a REAL  chance of causing the replay driver to fail. (And sometimes when they get in a bad state, they do not recover).

Reply:
Done. Just wondering if that makes the results unclear to users, since the behavior is quite different based on whether you're using direct or indirect draw calls.

Reply:
I agree, but crashing the replay would be even worse. We should absolutely document all of this. Alternatively, we can just leave indirect calls alone entirely.